feat(runs_on): permit using array of string to select specific runner

NikitaCOEUR opened the pull request:
## Check List

<!-- Please check the list. Please don't remove the check list. -->

- [X ] Read [CONTRIBUTING.md](https://github.com/suzuki-shunsuke/tfaction/blob/main/CONTRIBUTING.md)
  - [X ] Read [OSS Contribution Guide](https://github.com/suzuki-shunsuke/oss-contribution-guide/blob/main/README.md)
- [X ] [Write a GitHub Issue before creating a Pull Request](https://github.com/suzuki-shunsuke/oss-contribution-guide/blob/main/README.md#create-an-issue-before-creating-a-pull-request)
  - Link to the issue: https://github.com/suzuki-shunsuke/tfaction/issues/1738
- [X ] [All commits are signed](https://github.com/suzuki-shunsuke/oss-contribution-guide/blob/main/docs/commit-signing.md)
  - This repository enables `Require signed commits`, so all commits must be signed
- [X ] [Avoid force push](https://github.com/suzuki-shunsuke/oss-contribution-guide?tab=readme-ov-file#dont-do-force-pushes-after-opening-pull-requests)
- [ X] Do only one thing in one Pull Request

<!-- Please write the description here -->
Permit using array of string to select GitHub runners on which run jobs.
Necessary to select self hosted runners.

suzuki-shunsuke commented:
Your commits aren't singned.

> [X ] [All commits are signed](https://github.com/suzuki-shunsuke/oss-contribution-guide/blob/main/docs/commit-signing.md)
> This repository enables Require signed commits, so all commits must be signed

NikitaCOEUR commented:
Hi,

I saw that we could sign with our SSH key. I thought that since I was using my SSH key to clone the repo, the associated commits were automatically signed. But indeed, a bit of configuration was missing...

I have redone my commits as indicated in the procedure and force-pushed these two rebased commits with a signature.

I would like to ensure that all the changes I have made work correctly before merging them. It is quite complicated to test due to all the internal references to `suzuki-shunsuke/tfaction@main`. I saw that we could use a GitHub Action to create the PR and point to it in our workflows?

suzuki-shunsuke commented:
@NikitaCOEUR You can test this pull request using the version `pr/1740`.

- https://github.com/suzuki-shunsuke/tfaction/pull/1740#issuecomment-2214098796

NikitaCOEUR commented:
> @NikitaCOEUR You can test this pull request using the version `pr/1740`.
> 
> - https://github.com/suzuki-shunsuke/tfaction/pull/1740#issuecomment-2214098796

Yes thanks I saw that but I couldn't find any time for that now. Maybe next week!

NikitaCOEUR commented:
Hi,

I've added a new commit to fix an issue I introduced where the runs_on field was no longer optional in certain configurations. This should restore the previous behavior.

Additionally, I've retested everything, but I still encounter an error that I don't understand:
![image](https://github.com/user-attachments/assets/8a65cf28-6552-46a9-a23c-20cffaadae3d)
```bash
 bash: /home/runner/work/_actions/suzuki-shunsuke/github-action-terraform-init/71673ab29266f67c4470773675e087e9a4dda3c5/main.sh: No such file or directory
```

Despite this, the changes I've made seem to have the desired effect:

content of my tfaction-root.yaml : 
```yaml
target_groups:

# Target group 1 with specific self hosted runner (specified via array of strings)
- working_directory: terraform/env-TF-ACTION-TEST/
  target: TF-ACTION-TEST/
  environment: TF-ACTION-TEST
  runs_on:
    - self-hosted
    - runners
  secrets: # GitHub Secrets
  - env_name: AWS_ACCESS_KEY_ID # Environment variable name
    secret_name: BACKEND_ACCESS_KEY_ID
  - env_name: AWS_SECRET_ACCESS_KEY # Environment variable name
    secret_name: BACKEND_SECRET_KEY
  - env_name: SOPS_AGE_KEY
    secret_name: SOPS_AGE_KEY

# Target group 2 with specified runners (with string)
- working_directory: terraform/env-TF-ACTION-TEST2/
  target: TF-ACTION-TEST2/
  environment: TF-ACTION-TEST
  runs_on: ubuntu-latest
  secrets: # GitHub Secrets
  - env_name: AWS_ACCESS_KEY_ID # Environment variable name
    secret_name: BACKEND_ACCESS_KEY_ID
  - env_name: AWS_SECRET_ACCESS_KEY # Environment variable name
    secret_name: BACKEND_SECRET_KEY
  - env_name: SOPS_AGE_KEY
    secret_name: SOPS_AGE_KEY

# Target group 3 without any runs_on specified
- working_directory: terraform/env-TF-ACTION-TEST3/
  target: TF-ACTION-TEST3/
  environment: TF-ACTION-TEST
  secrets: # GitHub Secrets
  - env_name: AWS_ACCESS_KEY_ID # Environment variable name
    secret_name: BACKEND_ACCESS_KEY_ID
  - env_name: AWS_SECRET_ACCESS_KEY # Environment  variable name
    secret_name: BACKEND_SECRET_KEY
  - env_name: SOPS_AGE_KEY
    secret_name: SOPS_AGE_KEY
```
### Target group 1 with specific self hosted runner (specified via array of strings)
![image](https://github.com/user-attachments/assets/fa61ecbc-34a2-4982-b07b-2d7273f154d1)

### Target group 2 with specified runners (with string)
![image](https://github.com/user-attachments/assets/b96cb0ba-2225-4598-83b3-18ca028f40c1)

### Target group 3 without any runs_on specified
![image](https://github.com/user-attachments/assets/0a03dcfe-7a3e-4b4a-8bc3-d557c168172f)


Could you please help me with this issue?

Thanks!

suzuki-shunsuke commented:
```
 bash: /home/runner/work/_actions/suzuki-shunsuke/github-action-terraform-init/71673ab29266f67c4470773675e087e9a4dda3c5/main.sh: No such file or directory
```

I'm not sure why this error occurred. `main.sh` should exist.

```yaml
    - run: bash ${{ github.action_path }}/main.sh
```

https://github.com/suzuki-shunsuke/github-action-terraform-init/blob/71673ab29266f67c4470773675e087e9a4dda3c5/action.yaml#L28

This error would occur if you run this action in a container, but I guess you didn't do that.

- https://github.com/actions/runner/issues/2185

NikitaCOEUR commented:
Hmm ..
Indeed, I am using containers to run my GitHub actions.
That seems to be the problem.

I just saw a workaround suggested by someone: https://github.com/actions/runner/issues/716#issuecomment-795238933

I'm not sure if that would make sense here?

suzuki-shunsuke commented:
Could you allow me to push commits to the feature branch?

e.g. https://github.com/aquaproj/aqua-registry/pull/24089

```
Maintainers are allowed to edit this pull request.
```

<img width="376" alt="image" src="https://github.com/user-attachments/assets/2b7bfedc-e4ce-4452-92e7-f707c69ed485">

NikitaCOEUR commented:
So, I had to transfer the ownership of the fork to my user account to be able to activate this option.

suzuki-shunsuke commented:
> So, I had to transfer the ownership of the fork to my user account to be able to activate this option.

Oh, really. I didn't know that.
[The document](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) seems not to refer that.

Anyway, thank you for your update.

suzuki-shunsuke commented:
a0f260e84fa76fc2a970a05dbbb7c0156953e6b5 Replaced `${{github.action_path}}` with `$GITHUB_ACTION_PATH`.

NikitaCOEUR commented:
Here's the translation:

---

The purpose of this pull request seems fine.

However, I have issues with the [Taskfile](https://taskfile.dev/) combined with the execution of my GitHub Actions in a container.

I use a Taskfile at the root of my project as a wrapper for Terraform commands. This is used when a task command is executed from the root or in one of the subfolders. But they have implemented a safeguard based on directory ownership; task ascends through parent directories and stops if it does not find a Taskfile or as soon as the owner of the current directory changes compared to the previous one(https://github.com/go-task/task/blob/573949573968cc7056b98e0fdbee1d145ad869a9/taskfile/taskfile.go#L149) ... And this is the case since the containers on GitHub Actions are executed with the root user...

Also, the repo is cloned on the runner, and the volume is mounted in the container. When accessing the repo's folders, the root folder of the repo belongs to the user running Docker, but the subfolders belong to the user of the container (root).

NikitaCOEUR commented:
It seems that a workaround exists : https://github.com/actions/checkout/issues/47#issuecomment-536109176

I'll test that tomorrow.

suzuki-shunsuke commented:
> The purpose of this pull request seems fine.

Thank you for testing. It looks good.
Seems like your issue has nothing to do with tfaction.

NikitaCOEUR commented:
> It seems that a workaround exists : [actions/checkout#47 (comment)](https://github.com/actions/checkout/issues/47#issuecomment-536109176)
> 
> I'll test that tomorrow.
Doesn't work )=

But I have a workaround for my problem with Taskfile, which I posted in an issue: https://github.com/go-task/task/issues/1683#issuecomment-2234287796

However, this problem appears again with the git command. I just opened an issue here: https://github.com/suzuki-shunsuke/github-action-terraform-init/issues/19

suzuki-shunsuke commented:
> However, this problem appears again with the git command. I just opened an issue here: https://github.com/suzuki-shunsuke/github-action-terraform-init/issues/19

The issue was solved.

- https://github.com/suzuki-shunsuke/github-action-terraform-init/issues/19#issuecomment-2235974391

So everything is fine, right?

NikitaCOEUR commented:
Everything works perfectly when workflows run inside a container on our self-hosted runners and on GitHub runners!
![image](https://github.com/user-attachments/assets/320f8c10-892f-4b9d-befa-3703a0473bf9)

Workflows also run smoothly when they are not inside a container!

Tested and approved by me!

Thanks for your help and support.

NikitaCOEUR commented:
Do you have some doc to update for this feature ?

suzuki-shunsuke commented:
> Do you have some doc to update for this feature ?

```console
~/repos/src/github.com/suzuki-shunsuke/tfaction-docs main                                                                             07:49:57
$ git grep runs_on
docs/config/tfaction-root-yaml.md:  runs_on: ubuntu-latest # default is "ubuntu-latest". This is useful to use GitHub Actions Self Hosted Runner for the specific provider
```

https://github.com/suzuki-shunsuke/tfaction-docs/blob/bfc28814d4acfaf8d535de76509affe052a34d8e/docs/config/tfaction-root-yaml.md?plain=1#L177

suzuki-shunsuke commented:
I'll update the document.

NikitaCOEUR commented:
I can't ensure non-regression without running within a container. I have an issue; I don't know if it's caused by the transition from `${{GitHub.action_path}}` to `$GITHUB_ACTION` or if it's on my end.

suzuki-shunsuke commented:
> I have an issue

Could you explain the detail of your issue?

suzuki-shunsuke commented:
> I can't ensure non-regression without running within a container.

I've already confirmed this pr worked fine without a container.

NikitaCOEUR commented:
> > I have an issue
> 
> Could you explain the detail of your issue?
> 
> 

No, it's on my end then; the problem only occurs when I run without a container. I get an exit code 2 when running Terraform through my wrapper. But it must be due to the context change. I use the action within a container anyway, so there's no problem for me.

suzuki-shunsuke commented on schema/tfaction-root.json:
```diff
@@ -398,12 +401,19 @@
       "description": "environment variables"
+    },
+    "RunsOn": {
+      "type": ["string", "array"],
```
I'm not familiar with JSON Schema, so I'm not sure if this is correct.
I'll look into.

NikitaCOEUR replied:
https://cswr.github.io/JsonSchema/spec/multiple_types/
It seems correct, no ?

suzuki-shunsuke replied:
📝 `type` can accept an array.

https://json-schema.org/draft/2020-12/json-schema-core

```
"type": ["object", "boolean"],
```

But I think `array` is ambiguous because the type of elements is unknown. 🤔

NikitaCOEUR replied:
Maybe this is better :
```
"RunsOn": {
  "oneOf": [
    {
      "type": "string"
    },
    {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    {
      "type": "null"
    }
  ],
  "description": "The type of runner that the job will run on"
}
```

suzuki-shunsuke replied:
https://json-schema.org/understanding-json-schema/reference/combining#anyOf

Should we use `anyOf`?

```json
    "RunsOn": {
      "anyOf": [
        {
          "type": "string"
        },
        {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      ],
      "description": "The type of runner that the job will run on"
    }
```

NikitaCOEUR replied:
After reading more, Anyof seems more appropriate. It's more like a OR. So yes definitely.

NikitaCOEUR replied:
But null type is necessary because we can let empty the property no?

suzuki-shunsuke replied:
If `runs_on` is unset, the type definition is unrelated. So I don't think we need to add `null`.
If we want to support the setting which sets `null` explictly, we need to add `null`.
But I don't think we need to support `null` explicitly.

```yaml
runs_on: null
```

suzuki-shunsuke replied:
Fixed. 7290811a4c4481dc2f16f68d19e5cb4f8f85e890

suzuki-shunsuke commented:
v1.5.2 is out 🎉
https://github.com/suzuki-shunsuke/tfaction/releases/tag/v1.5.2
